activerecord/lib/active_record/persistence.rb

```diff
@@ -1154,6 +1157,16 @@ def apply_scoping?(options)
         (self.class.default_scopes?(all_queries: true) || self.class.global_current_scope)
     end
 
+    def with_scoping_rules(options, &block)
+      if apply_scoping?(options)
+        yield true
+      else
+        self.class.unscoped do
+          yield false
+        end
+      end
+    end
```

Comment:
This refactor was not strictly necessary, but I figured it would help sharing the same logic between `#reload` and `#increment!`